### PR TITLE
rust: find nested packages in git repositories

### PR DESCRIPTION
**Pull request by steveej**
rust: find nested packages in git repositories

this mimics the heuristic cargo uses for finding crates in their git
repositories ([cargo-issue-1462]).

[cargo-issue-1462]: https://github.com/rust-lang/cargo/issues/1462

###### Motivation for this change
`buildRustPackage` fails when used with the `cargoLock` argument and the package in the git dependency is not found in the top-level _Cargo.toml_. An example would this dependency specification:

```toml
cargo-test-support = { git = "https://github.com/rust-lang/cargo", branch = "rust-1.53.0" }
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by steveeJ** [comment]
@andir the test case is in place.

File: pkgs/build-support/rust/import-cargo-lock.nix [hunk 1]
```diff
@@ -125,20 +125,31 @@ let
...
-            ${jq}/bin/jq -r '.packages[] | select(.name == "${pkg.name}") | .manifest_path')
 
+        # If the target package is in a workspace, or if it's the top-level
+        # crate, we should find the crate path using `cargo metadata`.
+        crateCargoTOML=$(${cargo}/bin/cargo metadata --format-version 1 --no-deps --manifest-path $tree/Cargo.toml | \
+          ${jq}/bin/jq -r '.packages[] | select(.name == "${pkg.name}") | .manifest_path')
+
+        # If the repository is not a workspace the package might be in a subdirectory.
+        if [[ -z $crateCargoTOML ]]; then
+          for manifest in $(find $tree -name "Cargo.toml"); do
```
Comment (danieldk) [inline pkgs/build-support/rust/import-cargo-lock.nix:136]: What is the search order in Cargo? I'd think that in such cases a breadth-first would be preferred (higher in the tree is better). Also, with a breadth first search, this could be combined with the earlier `cargo metadata` call.

tl;dr: do you know what the traversal order in Cargo is? We probably want to do the same.

Comment (steveeJ) [reply pkgs/build-support/rust/import-cargo-lock.nix:136]: `cargo metadata` uses the [provided workspace member globs](https://github.com/rust-lang/cargo/blob/94ca096afbf25f670e76e07dca754fcfe27134be/src/cargo/core/workspace.rs#L1603-L1629) in the order they were given with the [globs expanding in alphabetical order](https://docs.rs/glob/0.3.0/glob/fn.glob.html).

since we are dealing with the case where the workspace members do not contain the desired crate, we can't viably use `cargo metadata` as an implementation guide. i don't have a strong opinion on the `find` order. both are predictable in knowing which crate will be found first in case there are multiple crates with the same name in the tree.

**Review by danieldk — APPROVED** [review]

File: pkgs/build-support/rust/test/import-cargo-lock/git-dependency-rev-non-workspace-nested-crate/Cargo.lock [hunk 2]
```diff
@@ -0,0 +1,638 @@
+# This file is automatically @generated by Cargo.
```
Comment (andir) [inline pkgs/build-support/rust/test/import-cargo-lock/git-dependency-rev-non-workspace-nested-crate/Cargo.lock:1]: This seems like a rather big example to test this case. Can't we build something more simple?

Comment (steveeJ) [reply pkgs/build-support/rust/test/import-cargo-lock/git-dependency-rev-non-workspace-nested-crate/Cargo.lock:1]: i didn't build the example, it's the real-world use-case that surfaced the bug for me. can you explain your concern further? i can't think of any practical downside.

Comment (andir) [reply pkgs/build-support/rust/test/import-cargo-lock/git-dependency-rev-non-workspace-nested-crate/Cargo.lock:1]: My only concern is complexity induced by the test case and runtime on hydra but those aren't really blockers.
